File: sections/35_math_mixed_contest/problems/m_totient_prefix/solution.py
import sys

MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    queries = data[1:1 + q]
    maximum = max(queries, default=1)

    phi = list(range(maximum + 1))
    for p in range(2, maximum + 1):
        if phi[p] != p:
            continue
        for multiple in range(p, maximum + 1, p):
            phi[multiple] -= phi[multiple] // p

    prefix = [0] * (maximum + 1)
    for value in range(1, maximum + 1):
        prefix[value] = (prefix[value - 1] + phi[value]) % MOD

    print("\n".join(str(prefix[n]) for n in queries))


if __name__ == "__main__":
    main()

File: sections/35_math_mixed_contest/problems/m_totient_prefix/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    cin >> q;
    vector<int> queries(q);
    int maximum = 1;
    for (int& n : queries) {
        cin >> n;
        maximum = max(maximum, n);
    }

    vector<int> phi(maximum + 1);
    iota(phi.begin(), phi.end(), 0);
    for (int p = 2; p <= maximum; ++p) {
        if (phi[p] != p)
            continue;
        for (int multiple = p; multiple <= maximum; multiple += p)
            phi[multiple] -= phi[multiple] / p;
    }

    vector<long long> prefix(maximum + 1);
    for (int i = 1; i <= maximum; ++i)
        prefix[i] = (prefix[i - 1] + phi[i]) % MOD;

    for (int n : queries)
        cout << prefix[n] << '\n';
}
